### ISSUE TITLE
Can't resolve plugins from plugin repository after Gradle 4.3 upgrade

### ISSUE THREAD

**Issue by edeandrea** [gradle/gradle#3326]
### Expected Behavior
When running my build, plugins declared in the `plugins` closure should be resolved from the Gradle Plugins Repository.

### Current Behavior
I get an error when my build starts up

### Context
All I did was upgrade from Gradle 4.2.1 to Gradle 4.3 and all of a sudden I am no longer able to run my build.

In Gradle 4.2.1 my build is fine, but with Gradle 4.3 my build fails with

```
FAILURE: Build failed with an exception.

* Where:
Build file '/Users/edeandre/workspaces/Stuff/my-gradle/build.gradle' line: 13

* What went wrong:
Error resolving plugin [id: 'com.jfrog.artifactory', version: '4.4.9']
> Could not GET 'https://plugins.gradle.org/api/gradle/4.3/plugin/use/com.jfrog.artifactory/4.4.9'.
   > Connect to plugins.gradle.org:443 [plugins.gradle.org/104.16.174.166, plugins.gradle.org/104.16.171.166, plugins.gradle.org/104.16.175.166, plugins.gradle.org/104.16.173.166, plugins.gradle.org/104.16.172.166, plugins.gradle.org/2400:cb00:2048:1:0:0:6810:afa6, plugins.gradle.org/2400:cb00:2048:1:0:0:6810:aba6, plugins.gradle.org/2400:cb00:2048:1:0:0:6810:aea6, plugins.gradle.org/2400:cb00:2048:1:0:0:6810:aca6, plugins.gradle.org/2400:cb00:2048:1:0:0:6810:ada6] failed: No route to host (connect failed)

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
```

This is my `plugins` block at the very top of my build.gradle:

```groovy
plugins {
	id 'java-gradle-plugin'
	id 'groovy'
	id 'maven-publish'
	id 'jacoco'
	id 'build-dashboard'
        id 'eclipse'
	id 'com.jfrog.artifactory' version '4.4.9'
	id "org.sonarqube" version "2.5"
}
```

I am not overriding anything to do with the plugin repositories. In fact, my project does not have a `settings.gradle` file.

**Comment by blindpirate** [comment]
This looks like a network issue. Does it still exist?

**Comment by edeandrea** [comment]
Yes it does. I can immediately switch back to Gradle 4.2.1 and everything is fine. As soon as I switch to 4.3 I experience this.

**Comment by blindpirate** [comment]
That's incredibly weird! We'll investigate it ASAP.

**Comment by edeandrea** [comment]
Thanks! I am behind a corporate proxy so not sure if there is something I need to do for https proxy that is new with 4.3?

**Comment by blindpirate** [comment]
AFAIK, no.

**Comment by blindpirate** [comment]
One more thing, do you always fail with `4.3` or intermittently? If this can be reproduced each time, could you please provide us with detailed stacktrace (run your build with `-s` again)? That would help a lot.

**Comment by edeandrea** [comment]
I will certainly do that for you in the morning. It always fails with 4.3.

**Comment by blindpirate** [comment]
Thanks!

**Comment by danlange** [comment]
I experienced a similar issue.  According to the 4.3 release notes, gradle no longer supports looking in multiple repositories for dependencies: https://docs.gradle.org/4.3/release-notes.html#avoid-checking-other-repositories-when-dependency-resolution-in-one-repository-fails

So, if your build.gradle includes a repository section with multiple repos, you will now get dependency resolution failures.  Looks like I'm stuck at 4.2.1.

**Comment by edeandrea** [comment]
For me my issue is with the plug-in repository. I have no settings.gradle so just the out of box defaults. 

That being said - if Gradle 4.3 doesn't allow dependencies to come from multiple repos then that will also be a show stopper for us. We absolutely require the ability to resolve from multiple repos. We are a global company and have many internal Artifactory servers distributed globally.

**Comment by jjohannes** [comment]
> if Gradle 4.3 doesn't allow dependencies to come from multiple repos then that will also be a show stopper for us. We absolutely require the ability to resolve from multiple repos. 

That's not what _dependency resolution in one repository fails_ means. I admit the wording could have been chosen better.
We fail if an actual connection error occurs (NOT when the dependency does not exist, which is not considered an error). `https://plugins.gradle.org/api/gradle/4.3/plugin/use/com.jfrog.artifactory/4.4.9` does exist. but there is a connection error when Gradle tries to retrieve it.

So I suspect that this is an error caused by your proxy configuration @edeandrea which was "hidden" before.

**Comment by FrancescoCapari** [comment]
This needs to be resolved ASAP, otherwise Gradle 4.3 is pretty much unusable.

After upgrading to 4.3, several of our Android projects are not able to resolve dependencies to libraries, mainly those from small developers that are in their custom repositories and not pushed to the big ones. 

Which should not be required as a change "Avoid checking other repositories when dependency resolution in one repository fails" forces to.

**TL;DR [WARNING] do not upgrade to Gradle 4.3 if you are using multiple repositories, otherwise you would not be able to build your projects.**

**Comment by jjohannes** [comment]
@FrancescoCapari thanks for the feedback. 

As stated above, Gradle _should_ not fail if a dependency does not exist at all in a repository. Only if the dependency exists but other errors occur. Of course there could be a problem that this behavior is not implemented correctly.

Can you share the error messages you are seeing? We need to find out what broke exactly in order to fix the problem.

**Comment by edeandrea** [comment]
Very weird - today I came in & can't seem to reproduce. I blew away my `~/.gradle/caches` folder & re-ran my build and everything seems to work fine. Very strange - I spent hours trying to figure out what was wrong yesterday...

**Comment by oehme** [comment]
Closing as the original issue can't be reproduced. 

@danlange @FrancescoCapari as Jendrik said, having multiple repositories is absolutely supported. The only change is that we fail if a repository returns a problem code like "forbidden" or "internal server error". If it returns "not found", we try the next one as usual.

Please check your proxy and authentication settings, your servers might be returning wrong status codes. If the problem persists, please open a new issue with reproduction instructions.

**Comment by archenroot** [comment]
I am having same issue on Gentoo with gradle 3.4.1:
```
unik /var/lib/layman/archenroot/app-misc/freeplane # eix gradle
[I] dev-java/gradle-bin
     Available versions:  
     (0)    (~)1.8[1]
     (3.3)  (~)3.3
     (3.4.1) (~)3.4.1
       {doc examples source ELIBC="FreeBSD"}
     Installed versions:  3.4.1(3.4.1)(04:09:57 AM 03/02/2018)(-doc -examples -source ELIBC="-FreeBSD")
     Homepage:            http://www.gradle.org/
     Description:         A project automation and build tool with a Groovy based DSL

```

Message:
```
> Configuring > 0/13 projects > root project > Compiling /var/tmp/portage/app-misc/freeplane-1.6.13/work/freeplane-1.6.13/build.gradle into local build cache > Compiling build file '/var/tmp/porta
FAILURE: Build failed with an exception.

* Where:
Build file '/var/tmp/portage/app-misc/freeplane-1.6.13/work/freeplane-1.6.13/build.gradle' line: 2

* What went wrong:
Error resolving plugin [id: 'nebula.ospackage', version: '4.4.0']
> Plugin cannot be resolved from https://plugins.gradle.org/api/gradle because Gradle is running in offline mode

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 3.005 secs
 * ERROR: app-misc/freeplane-1.6.13::archenroot failed (compile phase):
 *   (no error message)
 * 
 * Call stack:
 *     ebuild.sh, line 124:  Called src_compile
 *   environment, line 3602:  Called die
 * The specific snippet of code:
 *       ${GRADLE} build -x check -x test || die
 * 
 * If you need support, post the output of `emerge --info '=app-misc/freeplane-1.6.13::archenroot'`,
 * the complete build log and the output of `emerge -pqv '=app-misc/freeplane-1.6.13::archenroot'`.
!!! When you file a bug report, please include the following information:
GENTOO_VM=  CLASSPATH="" JAVA_HOME="/etc/java-config-2/current-system-vm"
JAVACFLAGS="" COMPILER=""
and of course, the output of emerge --info =freeplane-1.6.13
 * The complete build log is located at '/var/tmp/portage/app-misc/freeplane-1.6.13/temp/build.log'.
 * The ebuild environment file is located at '/var/tmp/portage/app-misc/freeplane-1.6.13/temp/environment'.
 * Working directory: '/var/tmp/portage/app-misc/freeplane-1.6.13/work/freeplane-1.6.13'
 * S: '/var/tmp/portage/app-misc/freeplane-1.6.13/work/freeplane-1.6.13'

```

**Comment by archenroot** [comment]
Otherwise I found the plugin in repo:
https://plugins.gradle.org/plugin/nebula.ospackage/4.4.0

**Comment by big-guy** [comment]
@archenroot That build is running with `--offline`, so it cannot build without having a local copy of the plugin somewhere (e.g., run once without `--offline`).

**Comment by jruuttun** [comment]
My first Gradle 4.6 build in a CI server behind corporate proxy failed with

```
Error resolving plugin [id: 'org.sonarqube', version: '2.6.2']
> Could not resolve all dependencies for configuration 'detachedConfiguration1'.
   > Could not determine artifacts for org.sonarqube:org.sonarqube.gradle.plugin:2.6.2
      > Could not get resource 'https://plugins.gradle.org/m2/org/sonarqube/org.sonarqube.gradle.plugin/2.6.2/org.sonarqube.gradle.plugin-2.6.2.jar'.
         > Could not HEAD 'https://plugins.gradle.org/m2/org/sonarqube/org.sonarqube.gradle.plugin/2.6.2/org.sonarqube.gradle.plugin-2.6.2.jar'. Received status code 403 from server: Forbidden
```
Downgrading to 4.2.1 fixed this particular error, but lead to another similar problem for another (transitive) build plugin dependency. Downgrading to 4.1 where I started fixed them both, possibly because of some previously created cache. I don't have control over the proxy configuration or the local ~/.gradle in the CI server, and I don't know why the 403 occurs in the first place. Local builds work fine with all of the aforementioned versions.

**Comment by jruuttun** [comment]
@eriwen thanks for your response. I'll try this when I can.

The resource url seems to redirect (303) to https://jcenter.bintray.com/org/sonarqube/org.sonarqube.gradle.plugin/2.6.2/org.sonarqube.gradle.plugin-2.6.2.jar, which in turn redirects (302) to https://repo.jfrog.org/artifactory/libs-release-bintray/org/sonarqube/org.sonarqube.gradle.plugin/2.6.2/org.sonarqube.gradle.plugin-2.6.2.jar?referrer, which responds with 404, with a json-formatted message "Could not find resource" . Indeed, the index at https://repo.jfrog.org/artifactory/libs-release-bintray/org/ doesn't list a *sonarqube/* subdirectory. 

This seems odd but doesn't explain the 403 response. Maybe the corporate proxy has whitelisted only some of those servers and not all of them? I feel that would explain the 403 but not the sudden breakage and remedy by downgrading, unless the plugin jars and dependencies come from some cache in the CI server.

**Comment by eriwen** [comment]
I cannot say what a proxy might be doing, but I imagine you can eliminate any local artifact cache from the scenario by using a custom `GRADLE_USER_HOME`. See 
https://docs.gradle.org/current/userguide/command_line_interface.html#_environment_options and https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_environment_variables for ways to configure this.

**Comment by eriwen** [comment]
@jruuttun Also, if this is something you really want to dig in and fix, you could also have your CI build create a [build scan](http://scans.gradle.com/) and you can check the network activity among other facets to debug this. Here's [an example](https://scans.gradle.com/s/tpqq7bqm74fwk/performance/networkActivity). It might give you a hint at what's really going on.

It sounds like you're in a corporate environment, so if you need to keep your project data private, you can run a [Gradle Enterprise trial](https://gradle.com/enterprise/trial) in Docker and that might help you out and not cost you money.

**Comment by jruuttun** [comment]
Adding the `-g [other than ~/.gradle]` option to the CI build revealed that the problem had indeed been hiding behind the local artifact cache for some time, and now broke the build even in 4.1. I still don't quite understand it but I should probably find a way of getting both in-house plugins *and* those applied using the plugins DSL resolved via the internal Nexus server, so that the build would not be dependent on whitelisting plugin source repositories in the beloved proxy. Thanks again for tips!

**Comment by eriwen** [comment]
@jruuttun The Gradle Plugin Portal is mirror-able, if that helps you in any way here.

**Comment by jruuttun** [comment]
Thanks. The problem went away after whitelisting repo.jfrog.org in the proxy.

I think the breakage after upgrading from 4.1 could have been a combined effect of #3799 and [this change](https://docs.gradle.org/4.3/release-notes.html#avoid-checking-other-repositories-when-dependency-resolution-in-one-repository-fails), making gradle try to reload all plugin artifacts and failing fast because of the proxy-originated 403.

**Comment by cgorgulla** [comment]
I have the same problem, I tried both Gratle 3.4.1 and 4.8.1, the problem seems to be the same.

Here is the issue I opened on Orchid: https://github.com/JavaEden/Orchid/issues/150#issuecomment-405328911

I don't use any proxy or VPN.

**Comment by oehme** [comment]
@cgorgulla That's a different issue and not caused by Gradle. There's some discussion about it on [StackOverflow](https://stackoverflow.com/questions/6784463/error-trustanchors-parameter-must-be-non-empty).

**Comment by ajeets1978** [comment]
To me this seems to be issue with JAVA8, where as when i use JAVA7 it works fine.

**Comment by rnnr** [comment]
Te me helped to add pluginManagement section to the start ot **settings.gradle**:

```
pluginManagement {
    repositories {
        maven { url "https://plugins.gradle.org/m2/" }
        gradlePluginPortal()
    }
}
```

**Comment by andrelimasantos** [comment]
> To me helped to add pluginManagement section to the start of the **settings.gradle** file:
> 
> ```
> pluginManagement {
>     repositories {
>         maven { url "https://plugins.gradle.org/m2/" }
>         gradlePluginPortal()
>     }
> }
> ```

Thanks!! That help me a lot!

**Comment by Shmarkus** [comment]
for me, the answer was this: https://spring.io/blog/2019/09/16/goodbye-http-repo-spring-use-https

**Comment by alexanderLinear** [comment]
for my best (but small) understanding
a build shall use all the specified repositories in the order given. (of course only for the use case they were intended by the specs.)

if one of the repositories fails to respond then this is sort of a fatal error because then there is the chance that some later order repository might provide a similar package (by name) that is different (by content). thus if not failing hard the build result might be different from what was requested (and expected) by the creating and invoking user.

**Comment by Dipesh199** [comment]
buildscript {
    dependencies {
        classpath "...."
    }
}


plugins {
    id 'com.android.application' version '7.1.1' apply false
    id 'com.android.library' version '7.1.1' apply false
    id 'org.jetbrains.kotlin.android' version '1.6.10' apply false


}

task clean(type: Delete) {
    delete rootProject.buildDir
}



Simply just write buildscript block and inside that block write dependencies block and inside that , you can add your classpath dependencies. 


Please dont make this issue so long by saying that , "I have same isssue or whatever...", 
Try to solve it. 

My project working fine with this method in new android versions gradle 4.2 also,,...

**Comment by omaraladin** [comment]
> @archenroot That build is running with `--offline`, so it cannot build without having a local copy of the plugin somewhere (e.g., run once without `--offline`).

This workarounds the problem actually for someone who's behind an enterprise proxy

**Comment by lockoutpython-appium** [comment]
I was challenged by that problem and discovered a way forward https://stackdev.space/questions/having-trouble-creating-a-liferay-project-in-eclipse-using-ant-ivy
